### PR TITLE
Support for building loaders/savers as dynamic loadable module

### PR DESCRIPTION
**Pull request by kleisauke**
This PR allows users to build a dynamically loadable module for a few dependencies as discussed in https://github.com/libvips/libvips/discussions/2018. This is enabled by default (i.e., `--with-{openslide,heif,poppler,magick}=module`) when:
* these dependencies are found;
* GModule is supported (`gmodule_supported` pkg-config variable).

And can be overridden on the command line with:
* `--without-{openslide,heif,poppler,magick}` - to disable OpenSlide/libheif/poppler/*magick usage;
* `--with-{openslide,heif,poppler,magick}[=yes]` - to restore the previous behavior;
* `--disable-modules` - to disable the build of dynamic modules.

This has been successfully tested on Linux (see https://github.com/weserv/rpms/commit/8f04e1300005ffab83547c3c7f1d2fd5c28b5ae6), Windows (see https://github.com/libvips/build-win64-mxe/commit/11d19eb17eb58911cd0237823883f7c4b180d029) and macOS (see GitHub actions build). Any further testing would be great!

Editor notes:
- The change set ended up being quite large due to the separation of the GObject and C part.
- ~The last 5 commits are not relevant to this PR. I could split this into separate PRs if needed.~ _Split into individual PRs_
- Modules built by autotools on macOS erroneously have the `.so` suffix instead of `.dylib`. See:  https://gitlab.gnome.org/GNOME/glib/-/issues/1413

Still to do:
- [X] ~Are `--enable-modules` and `--enable-static --disable-shared` mutually exclusive (i.e., should we disable the build of dynamic loadable modules when a statically linked library is built)?~ _A user may specify `--disable-modules` if desired, we should not force this setting._
- [X] Should we remove the use of the `.plg` suffix for modules? ~I still included this for backwards compatibility reasons.~ _Deprecated and removed when compiling with `--disable-deprecated`._
- [X] Are there any other dependencies that could benefit from this? _Added poppler and *magick as requested._ 
- [X] Rename `lib/vips-plugins-[MAJOR].[MINOR]` to `lib/vips-modules-[MAJOR].[MINOR]`? _Renaming done._

### CODE REVIEW

**Comment by remicollet** [comment]
Very nice WIP.

Quick test, seems to work pretty well.

Just a side note

```
$ ldd /usr/local/lib/vips-plugins-8.11/vips-heif.so  | grep heif
	libheif.so.1 => /lib64/libheif.so.1 (0x00007f5d0a487000)

```
Which is obviously expected

`$ ldd /usr/local/lib/libvips.so.42.13.0  | grep hei`

Which is also expected

BUT

```
$ ldd /usr/local/lib/libvips-cpp.so.42.13.0 | grep hei
	libheif.so.1 => /lib64/libheif.so.1 (0x00007f7f66ae3000)

```

Which is unwanted

**Comment by kleisauke** [comment]
Oh curious, on which distro was that tested on? I couldn't reproduce this on Fedora 33. I remembered that libtool also links against the dependencies of `libvips.so.42.13.0`, see for example:
<details>
  <summary>Details</summary>

```bash
$ readelf -d /usr/local/lib/vips-plugins-8.11/vips-heif.so | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libvips.so.42]
 0x0000000000000001 (NEEDED)             Shared library: [libexpat.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libgsf-1.so.114]
 0x0000000000000001 (NEEDED)             Shared library: [libxml2.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libfftw3.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libMagickCore-6.Q16.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [liborc-0.4.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [liblcms2.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libIlmImf-2_3.so.24]
 0x0000000000000001 (NEEDED)             Shared library: [libImath-2_3.so.24]
 0x0000000000000001 (NEEDED)             Shared library: [libHalf.so.24]
 0x0000000000000001 (NEEDED)             Shared library: [libIex-2_3.so.24]
 0x0000000000000001 (NEEDED)             Shared library: [libIexMath-2_3.so.24]
 0x0000000000000001 (NEEDED)             Shared library: [libIlmThread-2_3.so.24]
 0x0000000000000001 (NEEDED)             Shared library: [libniftiio.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libznz.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libopenjp2.so.7]
 0x0000000000000001 (NEEDED)             Shared library: [libpoppler-glib.so.8]
 0x0000000000000001 (NEEDED)             Shared library: [librsvg-2.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libgdk_pixbuf-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libcairo.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libmatio.so.9]
 0x0000000000000001 (NEEDED)             Shared library: [libhdf5.so.103]
 0x0000000000000001 (NEEDED)             Shared library: [libcfitsio.so.7]
 0x0000000000000001 (NEEDED)             Shared library: [libwebpmux.so.3]
 0x0000000000000001 (NEEDED)             Shared library: [libwebpdemux.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libwebp.so.7]
 0x0000000000000001 (NEEDED)             Shared library: [libpangoft2-1.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libpango-1.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libharfbuzz.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libfontconfig.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libfreetype.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libtiff.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libspng.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libpng16.so.16]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libimagequant.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libjpeg.so.62]
 0x0000000000000001 (NEEDED)             Shared library: [libexif.so.12]
 0x0000000000000001 (NEEDED)             Shared library: [libgmodule-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgio-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgobject-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libglib-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libheif.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libaom.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libde265.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libx265.so.192]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```
</details>

This can be circumvented when doing:
```bash
sed -i '/newdependency_libs="/s/$deplib //' libtool
```
(see https://lists.gnu.org/archive/html/libtool/2005-12/msg00021.html)

Just after configuring libvips. For example, with that I see:
<details>
  <summary>Details</summary>

```bash
$ readelf -d /usr/local/lib/vips-plugins-8.11/vips-heif.so | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libvips.so.42]
 0x0000000000000001 (NEEDED)             Shared library: [libgmodule-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgio-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgobject-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libglib-2.0.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libheif.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libaom.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libde265.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libx265.so.192]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```
</details>

**Comment by kleisauke** [comment]
I just noticed that this doesn't work out-of-the-box when libvips is installed in a non-standard prefix. See:
```bash
$ /usr/local/bin/vips heifload
vips: unknown action "heifload"
$ strace /usr/local/bin/vips heifload 2>&1 | grep plugins
openat(AT_FDCWD, "/usr/lib/vips-plugins-8.11", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
$ VIPSHOME=/usr/local /usr/local/bin/vips heifload
load a HEIF image
...
```

**Comment by remicollet** [comment]
>  Are there any other dependencies that could benefit from this?

Possibly ImageMagick

As 2 major versions exists, it could be interesting to build the module twice and distribute them separately to give more choice to users... but in all case, IM is a nightmare...

**Comment by jcupitt** [comment]
Poppler could benefit too --- it's GPL, so some users would prefer not to have it linked in.

**Comment by remicollet** [comment]
> Oh curious, on which distro was that tested on?

Sorry for the noise , probably some build artefact from a non clean env.
(btw, Fedora 33 for now)

New test this morning, and everything seems OK.

```
$ ./configure --prefix=/opt/libvips --libdir=/opt/libvips/lib64
$ make && make install
$ /opt/libvips/bin/vips heifload
load a HEIF image
...
$ ldd /opt/libvips/lib64/*so | grep sli
$ ldd /opt/libvips/lib64/*so | grep hei
$ ldd /opt/libvips/lib64/vips-plugins-8.11/*so | grep hei
/opt/libvips/lib64/vips-plugins-8.11/vips-heif.so:
	libheif.so.1 => /lib64/libheif.so.1 (0x00007f29747c2000)
$ ldd /opt/libvips/lib64/vips-plugins-8.11/*so | grep sli
/opt/libvips/lib64/vips-plugins-8.11/vips-openslide.so:
	libopenslide.so.0 => /lib64/libopenslide.so.0 (0x00007f1165b4c000)
$ rm -rf /opt/libvips/lib64/vips-plugins-8.11
$ /opt/libvips/bin/vips heifload
vips: unknown action "heifload"

```

> I just noticed that this doesn't work out-of-the-box when libvips is installed in a non-standard prefix. See:

No issue for me

**Comment by kleisauke** [comment]
OK, added Poppler and *Magick as requested. I also resolved a couple of TODO-items, this PR is now ready for review.

> No issue for me

It seems I still had `VIPSHOME=/usr` set in my bash profile, sorry for the noise.

**Comment by jcupitt** [comment]
This is very cool, and works well for me! Wonderful work.

I noticed that `vips.pc` still lists the module packages under `Requires.private` even though they are no longer required for linking. We could safely remove them (I think).

I've no idea how this change would interact with static linking, though I think that's barely supported anyway.

I'll keep reading.

**Comment by kleisauke** [comment]
Ah yes, you're right. I've removed the module packages from `Requires.private`.

I suppose it should work when libvips is built statically, but I haven't tried that yet. Perhaps it would be possible to dynamically link Poppler within the loadable module while libvips (and the other non-GPL dependencies) are statically built? This could help users avoid violating the GPL license.

**Comment by jcupitt** [comment]
Yes, I think so, as long as libglibc is dynamic I think you can still dlopen() shared libraries.

**Comment by jcupitt** [comment]
OK, all read, it looks good, let's merge!

This is a nice thing.

File: libvips/foreign/heif2vips.c [hunk 1]
```diff
@@ -0,0 +1,1267 @@
+/* load heif images with libheif
```
Comment (jcupitt) [inline libvips/foreign/heif2vips.c:1]: How about moving the definition of `heifload` back into `heifload.c`, and putting the C API wrappers for all these modules into `foreign.c`?

Comment (kleisauke) [reply libvips/foreign/heif2vips.c:1]: Ah, yes, that is definitely better. I'm not sure why I didn't think of this in the first place. :sweat_smile:

Comment (jcupitt) [reply libvips/foreign/heif2vips.c:1]: OK, done!

**Pull request merged**